openedx/core/djangoapps/xblock/README.rst

```diff
@@ -0,0 +1,94 @@
+   - Avoiding XModule support as much as possible keeps the code cleaner
+   - All XModule compatibility code that we do include must be isolated within the ``shims.py`` file for eventual removal
+   - As a firm rule, any block with code split into separate ``Module`` and ``Descriptor`` classes will not be supported.
+* To support learning experiences other than courses
+   - The new runtime is built around the concept of "learning contexts," which are more generic and flexible than courses. See `Learning Contexts`_ below.
+   - Additional learning contexts and related functionality can be added via plugins.
+   - The new runtime makes as few assumptions as possible about how XBlocks are structured (doesn't expect a rooted tree hierarchy like every course and library in the old runtime must have).
+* To avoid any frontend code
+   - This new runtime is API-first, designed to be consumed by SPA frontends or mobile apps rather than rendering HTML into templates and sending them to the browser. Since it is based on XBlocks, there are some circumstances where it returns HTML generated by the XBlock views themselves, but it returns that HTML as an API response only, wrapped in a JSON "Fragment" object. In no other circumstances does this new runtime produce HTML.
+   - This is in contrast to the old runtime in which key parts of the LMS UI were added to HTML output by the runtime and/or the XBlocks themselves (e.g. the ``sequential`` XBlock rendered next/prev navigation buttons).
+* To provide first-class support for anonymous usage of XBlocks (use by unregistered users)
+   - If the user is unregistered, field data will be saved into the user's django session rather than the database.
+* To support sandboxed execution of XBlocks
+   - The runtime API provides a mechanism for calling XBlock handlers without session authentication (using a secure token in the URL), so that XBlocks can be run in sandboxed IFrames that cannot access the user's LMS cookies or interact with the LMS other than through the XBlock JavaScript API.
+
+Code Layout and Separation of Concerns
+--------------------------------------
+
+All of this code is found in this ``openedx.core.djangoapps.xblock`` folder:
+
+* `api.py <./api.py>`_ : **Python API** for this new XBlock runtime. Provides methods for loading and interacting with XBlocks given a user and a usage ID, such as: ``load_block()``, ``get_block_metadata()``, ``render_block_view()``, and ``get_handler_url()``
+
+  - Implementation differences in how Studio and the LMS implement these APIs are encapsulated in ``lms_impl.py`` and ``studio_impl.py``
+* `rest_api <./rest_api/>`_ : **REST API** for this new XBlock runtime. Provides `endpoints <./rest_api/urls.py>`_ for things like getting an XBlock's metadata, rendering an XBlock's view, calling an XBlock's handler, etc.
+  - An example URL is https://studio.example.com/api/xblock/v1/xblocks/lb:library1:html:introduction/view/student_view/
```

Comment:
Don't we have an /api/xblock endpoint already? Or is that just /xblock? In any case, does it make sense to use a different top level identifier? Or maybe a v2 indicator?

Reply:
The equivalent JSON XBlock API to this is at
`/courses/{course_key}/xblock/{usage_key}/view/{view_name}` and
`/courses/{course_key}/xblock/{usage_key}/handler/{handler_name}`.
You're right that there is also an endpoint at `/xblock/{usage_key}` but it's not an API in the same sense because it renders the "chromeless" view of the XBlock (for mobile apps), and includes a lot of LMS CSS + JS in addition to the XBlock HTML.

In any case, I think using "v1" is a good idea, even though there's no "v1" under the same URL format, because it will otherwise be fairly confusing as to which URLs are which.

Reply:
Changed to `v2`.